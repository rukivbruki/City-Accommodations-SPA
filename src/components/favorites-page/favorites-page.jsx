import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {compose} from 'recompose';

import FavoritesCard from '../favorites-card/favorites-card.jsx';
import Spinner from '../spinner/spinner.jsx';
import Footer from '../footer/footer.jsx';

import {loadFavorites} from '../../reducers/favorites/favorites';
import {getIsLoaded, getLoadingError, getOrderedFavorites} from '../../reducers/favorites/selectors';
import withAuthorizationReqired from '../../hocs/with-authorization-reqired/with-authorization-reqired';
import {setBodyRootClass} from "../../utils/set-body-root-class";

class FavoritePage extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.loadData();
    setBodyRootClass(`page`, ``);
  }

  render() {
    const {favorites, isLoaded, loadingError} = this.props;

    if (!isLoaded) {
      return <>
        <main className="page__main page__main--favorites page__main--favorites-empty">
          <div className="page__favorites-container container">
            <Spinner />
          </div>
        </main>
        <Footer/>
      </>;
    }

    if (!favorites || !favorites.length) {
      return <>
        <main className="page__main page__main--favorites page__main--favorites-empty">
          <div className="page__favorites-container container">
            <section className="favorites favorites--empty">
              <h1 className="visually-hidden">Favorites (empty)</h1>
              <div className="favorites__status-wrapper">
                <b className="favorites__status">Nothing yet saved.</b>
                <p className="favorites__status-description">Save properties to narrow
                  down search or plan yor future trips.</p>
              </div>
            </section>
          </div>
        </main>
        <Footer/>
      </>;
    }

    if (loadingError) {
      return <>
        <main className="page__main page__main--favorites page__main--favorites-empty">
          <div className="page__favorites-container container">
            <section className="favorites favorites--empty">
              <h1 className="visually-hidden">Load favorites error: {loadingError}</h1>
            </section>
          </div>
        </main>
        <Footer/>
      </>;
    }

    return <>
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {favorites.map((it) => {
                return <li key={it.city.name} className="favorites__locations-items">
                  <div className="favorites__locations locations locations--current">
                    <div className="locations__item">
                      <a className="locations__item-link" href="#">
                        <span>{it.city.name}</span>
                      </a>
                    </div>
                  </div>
                  <div className="favorites__places">
                    {it.offers.map((offer) => {
                      return <FavoritesCard place={offer} key={offer.id}/>;
                    })}
                  </div>
                </li>;
              })}
            </ul>
          </section>
        </div>
      </main>
      <Footer/>
    </>;
  }
}

FavoritePage.propTypes = {
  loadData: PropTypes.func.isRequired,
  isLoaded: PropTypes.bool.isRequired,
  loadingError: PropTypes.string,
  favorites: PropTypes.arrayOf(PropTypes.shape({
    city: PropTypes.shape({
      name: PropTypes.string.isRequired
    }),
    offers: PropTypes.array,
  })),
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  favorites: getOrderedFavorites(state),
  isLoaded: getIsLoaded(state),
  loadError: getLoadingError(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadData: () => dispatch(loadFavorites())
});

export {FavoritePage};
export default compose(
    connect(mapStateToProps, mapDispatchToProps),
    withAuthorizationReqired
)(FavoritePage);
