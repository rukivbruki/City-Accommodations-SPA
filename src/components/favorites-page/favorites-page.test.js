import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {FavoritePage} from './favorites-page';
import mockPlaces from '../../mocks/mock-offers';

jest.mock(`../../utils/set-body-root-class`);

it(`FavoritePage snapshot`, () => {
  const renderer = new ShallowRenderer();
  const result = renderer.render(<FavoritePage
    loadData={jest.fn()}
    isLoaded={true}
    loadingError={null}
    favorites={[{
      city: {name: `city`},
      offers: mockPlaces
    }]}
  />);
  expect(result).toMatchSnapshot();
});

