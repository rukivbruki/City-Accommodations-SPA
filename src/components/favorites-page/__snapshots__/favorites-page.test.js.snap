// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritePage snapshot 1`] = `
<React.Fragment>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    city
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <Memo(ConnectFunction)
                place={
                  Object {
                    "bedrooms": 2,
                    "city": Object {
                      "location": Object {
                        "latitude": 1,
                        "longitude": 2,
                        "zoom": 13,
                      },
                      "name": "City name 1",
                    },
                    "description": "text",
                    "goods": Array [
                      "one",
                      "two",
                      "tree",
                    ],
                    "host": Object {
                      "avatarUrl": "url",
                      "id": 1,
                      "isPro": true,
                      "name": "name",
                    },
                    "id": 999,
                    "images": Array [
                      "url1",
                      "url2",
                      "url3",
                    ],
                    "isFavorite": true,
                    "isPremium": true,
                    "location": Object {
                      "latitude": 1,
                      "longitude": 2,
                      "zoom": 13,
                    },
                    "maxAdults": 2,
                    "picture": "test-image1.jpg",
                    "previewImage": "",
                    "price": 111,
                    "rating": 22,
                    "title": "Test title 1",
                    "type": "apartment",
                  }
                }
              />
              <Memo(ConnectFunction)
                place={
                  Object {
                    "bedrooms": 3,
                    "city": Object {
                      "location": Object {
                        "latitude": 1,
                        "longitude": 2,
                        "zoom": 13,
                      },
                      "name": "City name 2",
                    },
                    "coordinates": Array [
                      52.3909553943508,
                      4.85309666406198,
                    ],
                    "description": "text",
                    "goods": Array [
                      "one",
                      "two",
                      "tree",
                    ],
                    "host": Object {
                      "avatarUrl": "url",
                      "id": 1,
                      "isPro": true,
                      "name": "name",
                    },
                    "id": 222,
                    "images": Array [
                      "url1",
                      "url2",
                      "url3",
                    ],
                    "isFavorite": true,
                    "isPremium": false,
                    "location": Object {
                      "latitude": 1,
                      "longitude": 2,
                      "zoom": 13,
                    },
                    "maxAdults": 1,
                    "picture": "test-image2.jpg",
                    "previewImage": "",
                    "price": 222,
                    "rating": 100,
                    "title": "Test title 2",
                    "type": "house",
                  }
                }
              />
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <Footer />
</React.Fragment>
`;
